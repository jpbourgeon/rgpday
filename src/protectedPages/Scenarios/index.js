import React from 'react'
import { Redirect } from '@reach/router'
import PropTypes from 'prop-types'
import { withStyles } from '@material-ui/core/styles'
import Paper from '@material-ui/core/Paper'
import Breadcrumbs from '@material-ui/lab/Breadcrumbs'
import Typography from '@material-ui/core/Typography'
import Grid from '@material-ui/core/Grid'
import Divider from '@material-ui/core/Divider'
import IconButton from '@material-ui/core/IconButton'
import Button from '@material-ui/core/Button'
import TextField from '@material-ui/core/TextField'
import DialogTitle from '@material-ui/core/DialogTitle'
import DialogContent from '@material-ui/core/DialogContent'
import DialogContentText from '@material-ui/core/DialogContentText'
import DialogActions from '@material-ui/core/DialogActions'
import Dialog from '@material-ui/core/Dialog'
import Fab from '@material-ui/core/Fab'
import Folder from '@material-ui/icons/FolderOpenOutlined'
import Search from '@material-ui/icons/Search'
import Close from '@material-ui/icons/Close'
import Delete from '@material-ui/icons/Delete'
import Edit from '@material-ui/icons/Edit'
import NavigateNext from '@material-ui/icons/NavigateNext'
import Add from '@material-ui/icons/Add'
import Link from 'src/components/Link'
import MUILink from '@material-ui/core/Link'
import Tile from 'src/components/Tile'
import { deleteScenario } from 'src/graphql/mutations'
import logger from 'src/logger'
import API, { graphqlOperation } from '@aws-amplify/api'
import config from 'src/aws-exports'

API.configure(config)
const listScenarios = `query ListScenarios(
  $filter: ModelScenarioFilterInput
  $limit: Int
  $nextToken: String
  ) {
    listScenarios(filter: $filter, limit: $limit, nextToken: $nextToken) {
      items {
        id
        description
        searchable
        sessions {
          items {
            id
          }
          nextToken
        }
      }
      nextToken
    }
  }
  `

const styles = theme => {
  return {
    '@global': {
      '.grecaptcha-badge': {
        visibility: 'hidden'
      }
    },
    layout: {
      width: 'auto',
      marginLeft: theme.spacing.unit * 3,
      marginRight: theme.spacing.unit * 3,
      [theme.breakpoints.up(1100 + theme.spacing.unit * 3 * 2)]: {
        width: 1100,
        marginLeft: 'auto',
        marginRight: 'auto'
      }
    },
    paper: {
      backgroundSize: `100% 100%`,
      marginBottom: theme.spacing.unit * 4,
      padding: theme.spacing.unit * 4,
      textAlign: 'justify'
    },
    breadcrumb: {
      marginBottom: theme.spacing.unit * 2
    },
    title: {
      marginBottom: theme.spacing.unit * 2
    },
    titleIcon: {
      fontSize: '2.5rem',
      marginBottom: '-0.5rem'
    },
    fab: {
      [theme.breakpoints.up('sm')]: {
        float: 'right'
      },
      [theme.breakpoints.down('xs')]: {
        position: 'fixed',
        bottom: theme.spacing.unit,
        right: theme.spacing.unit
      }
    },
    grow: {
      flexGrow: 1
    },
    itemActionIcon: {
      width: '1rem',
      height: '1rem'
    },
    loadMoreLink: {
      cursor: 'pointer'
    },
    loadMoreItem: {
      textAlign: 'center'
    },
    hide: {
      display: 'none'
    },
    formContainer: {
      marginBottom: theme.spacing.unit * 2
    },
    filterButtonsContainer: {
      display: 'inline-flex',
      flexDirection: 'column',
      textAlign: 'center',
      justifyContent: 'center'
    },
    filterButtons: {
      padding: theme.spacing.unit,
      marginTop: theme.spacing.unit
    },
    dialog: {
      padding: '1em'
    }
  }
}

const defaultState = {
  items: [],
  limit: 10,
  nextToken: null,
  filter: '',
  delete: {
    id: '',
    openDialog: false
  }
}

class Component extends React.Component {
  constructor (props) {
    super(props)
    this._isMounted = false
    this.state = defaultState
    this.loadItems = this.loadItems.bind(this)
    this.handleChangeFilter = this.handleChangeFilter.bind(this)
    this.handleCancelFilter = this.handleCancelFilter.bind(this)
    this.handleDeleteDialog = this.handleDeleteDialog.bind(this)
    this.deleteScenario = this.deleteScenario.bind(this)
  }

  componentDidMount () {
    this._isMounted = true
    this.loadItems()
  }

  componentWillUnmount () {
    this._isMounted = false
  }

  async loadItems (event = { preventDefault: () => {} }) {
    event.preventDefault()
    try {
      const state = this.state
      if (state.nextToken === null) state.items = []
      let filter = null
      if (this.state.filter !== '') {
        filter = { searchable: { contains: this.state.filter.toLowerCase() } }
      }
      // GraphQL
      const result = await API.graphql(
        graphqlOperation(listScenarios, {
          filter,
          limit: this.state.limit,
          nextToken: this.state.nextToken
        })
      )
      if (!result.errors) {
        state.nextToken = result.data.listScenarios.nextToken
        state.items = [...state.items, ...result.data.listScenarios.items]
        if (this._isMounted) {
          this.setState(state, () => {
            if (result.data.listScenarios.items.length === 0 && result.data.listScenarios.nextToken !== null) {
              this.loadItems()
            }
          })
        }
      } else {
        logger.error('loadItems', result)
      }
    } catch (error) {
      logger.error('loadItems', error)
    }
  }

  handleChangeFilter (event) {
    const state = this.state
    state.filter = event.target.value
    this.setState(state)
  }

  handleCancelFilter (event) {
    event.preventDefault()
    this.setState(defaultState, () => {
      this.loadItems()
    })
  }

  handleDeleteDialog (id, openDialog) {
    this.setState({
      delete: {
        id,
        openDialog
      }
    })
  }

  async deleteScenario () {
    try {
      // GraphQL
      const id = this.state.delete.id
      this.handleDeleteDialog('', false)
      const result = await API.graphql(
        graphqlOperation(deleteScenario, { input: { id } })
      )
      if (!result.errors) {
        const items = this.state.items
        const index = items.findIndex(item => item.id === id)
        if (index > -1) {
          items.splice(index, 1)
        }
        this.setState({ items })
      } else {
        logger.error('deleteScenario', result)
      }
    } catch (error) {
      logger.error('deleteScenario', error)
    }
  }

  render () {
    const { classes, config } = this.props
    if (!config.isAdmin) return (<Redirect noThrow to='/dashboard' />)
    const renderActions = (id) => (
      <Grid container>
        <Grid item className={classes.grow}>{id}</Grid>
        <Grid item>
          <Link to={`./update/${id}`}>
            <IconButton><Edit className={classes.itemActionIcon} /></IconButton>
          </Link>
          <MUILink onClick={() => { this.handleDeleteDialog(id, true) }}>
            <IconButton><Delete size='small' className={classes.itemActionIcon} /></IconButton>
          </MUILink>
        </Grid>
      </Grid>
    )
    const renderItems = (items) => {
      return (
        items
          .map((item) => {
            return (
              <Tile
                key={item.id}
                title={renderActions(item.id)}
                description={(
                  <React.Fragment>
                    <span>{item.description.split('\n').map((line, key) => {
                      return (<React.Fragment key={key}>{line}<br /></React.Fragment>)
                    })}</span>
                    {(!(Array.isArray(item.sessions.items) && item.sessions.items.length > 0)) ? null : (
                      <React.Fragment>
                        <br /><span><strong>Sessions : </strong>{item.sessions.items.map(i => {
                          return <Link key={i.id} to={`../sessions/update/${i.id}`}>{i.id}</Link>
                        })}</span>
                      </React.Fragment>
                    )}
                  </React.Fragment>
                )}
              />
            )
          })
      )
    }
    return (
      <div className={classes.layout}>
        <main>
          <Paper className={classes.paper}>
            <Grid container className={classes.container}>
              <Grid item xs={12} className={classes.breadcrumb}>
                <Breadcrumbs separator={<NavigateNext fontSize='small' />} arial-label='Breadcrumb'>
                  <Link to='/' color='inherit'>Accueil</Link>
                  <Link to='../' color='inherit'>Tableau de bord</Link>
                  <Link to='.' color='inherit'>Scénarios</Link>
                </Breadcrumbs>
              </Grid>
              <Grid item xs={12} className={classes.title}>
                <Typography variant='h4' gutterBottom>
                  <div className={classes.fab}>
                    <Link to='./add'>
                      <Fab color='primary' aria-label='Ajouter'>
                        <Add />
                      </Fab>
                    </Link>
                  </div>
                  <Folder className={classes.titleIcon} />&nbsp;Scénarios
                </Typography>
                <Divider />
              </Grid>
              <Grid item xs={12}>
                <form
                  className={classes.formContainer}
                  noValidate
                  autoComplete='off'
                  onSubmit={(e) => { this.loadItems(e, true) }}
                >
                  <Grid container>
                    <Grid item className={classes.grow}>
                      <TextField
                        label='Filtrer'
                        fullWidth
                        margin='normal'
                        variant='outlined'
                        onChange={(e) => this.handleChangeFilter(e)}
                        value={this.state.filter}
                      />
                    </Grid>
                    <Grid item className={classes.filterButtonsContainer}>
                      <div className={classes.filterButtons}>
                        <IconButton
                          type='submit'
                        >
                          <Search />
                        </IconButton>
                        <IconButton
                          onClick={this.handleCancelFilter}
                        >
                          <Close />
                        </IconButton>
                      </div>
                    </Grid>
                  </Grid>
                </form>
              </Grid>
              <Grid item xs={12}>
                <Grid container spacing={40}>
                  {(this.state.items.length > 0) ? renderItems(this.state.items) : (
                    <Grid item>
                      <Typography variant='body1' gutterBottom>Il n'y a pas de scénario à afficher</Typography>
                    </Grid>
                  )}
                  <Grid item xs={12} className={(this.state.nextToken === null) ? classes.hide : null}>
                    <Typography variant='body1' gutterBottom className={classes.loadMoreItem}>
                      <MUILink
                        onClick={this.loadItems}
                        className={classes.loadMoreLink}>
                          Afficher plus de résultats
                      </MUILink>
                    </Typography>
                  </Grid>
                </Grid>
              </Grid>
            </Grid>
          </Paper>
        </main>
        <Dialog
          open={this.state.delete.openDialog}
          onClose={() => { this.handleDeleteDialog('', false) }}
          aria-labelledby='alert-dialog-title'
          aria-describedby='alert-dialog-description'
        >
          <DialogTitle id='alert-dialog-title'>Effacer le scénario {this.state.delete.id}</DialogTitle>
          <DialogContent>
            <DialogContentText id='alert-dialog-description'>
              Attention, cette action est irréversible.
            </DialogContentText>
          </DialogContent>
          <DialogActions className={classes.dialog}>
            <Button onClick={this.deleteScenario} color='secondary' variant='contained'>
              Confirmer
            </Button>
            <Button onClick={() => { this.handleDeleteDialog('', false) }} color='primary' autoFocus variant='outlined'>
              Annuler
            </Button>
          </DialogActions>
        </Dialog>
      </div>
    )
  }
}

Component.propTypes = {
  classes: PropTypes.object.isRequired
}

export default withStyles(styles)(Component)
