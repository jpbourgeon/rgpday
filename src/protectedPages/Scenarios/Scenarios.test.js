/* global describe, it, expect */
import React from 'react'
import renderer from 'react-test-renderer'
import Scenarios from './'
import jssSerializer from 'src/utils/jssSnapshotSerializer'

expect.addSnapshotSerializer(jssSerializer)

describe('The Scenarios component', () => {
  it('should render correctly', () => {
    const tree = renderer.create(<Scenarios />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
