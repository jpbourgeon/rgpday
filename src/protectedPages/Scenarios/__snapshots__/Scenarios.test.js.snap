// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Scenarios component should render correctly 1`] = `
<div
  className="layout-1"
>
  <main>
    <div
      className="root-16 elevation2-20 rounded-17 paper-2"
    >
      <div
        className="container-43"
      >
        <div
          className="item-44 12-83 breadcrumb-3"
        >
          <nav
            arial-label="Breadcrumb"
            className="root-144 body1-153 colorTextSecondary-177 root-140"
          >
            <ol
              className="ol-141"
            >
              <li
                className="li-142"
              >
                <a
                  aria-current="page"
                  className="root-144 colorInherit-173 inline-179 root-180 underlineHover-182"
                  href="/"
                  onClick={[Function]}
                >
                  Accueil
                </a>
              </li>
              <li
                aria-hidden="true"
                className="root-185 separator-143"
              >
                <svg
                  aria-hidden="true"
                  className="root-186 fontSizeSmall-193"
                  focusable="false"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                  />
                  <path
                    d="M0 0h24v24H0z"
                    fill="none"
                  />
                </svg>
              </li>
              <li
                className="li-142"
              >
                <a
                  aria-current="page"
                  className="root-144 colorInherit-173 inline-179 root-180 underlineHover-182"
                  href="/"
                  onClick={[Function]}
                >
                  Tableau de bord
                </a>
              </li>
              <li
                aria-hidden="true"
                className="root-185 separator-143"
              >
                <svg
                  aria-hidden="true"
                  className="root-186 fontSizeSmall-193"
                  focusable="false"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                  />
                  <path
                    d="M0 0h24v24H0z"
                    fill="none"
                  />
                </svg>
              </li>
              <li
                className="li-142"
              >
                <a
                  aria-current="page"
                  className="root-144 colorInherit-173 inline-179 root-180 underlineHover-182"
                  href="/"
                  onClick={[Function]}
                >
                  Scénarios
                </a>
              </li>
            </ol>
          </nav>
        </div>
        <div
          className="item-44 12-83 title-4"
        >
          <h4
            className="root-144 h4-159 gutterBottom-171"
          >
            <div
              className="fab-6"
            >
              <a
                className="root-144 colorPrimary-174 inline-179 root-180 underlineHover-182"
                href="//add"
                onClick={[Function]}
              >
                <button
                  aria-label="Ajouter"
                  className="root-205 root-195 primary-197"
                  disabled={false}
                  onBlur={[Function]}
                  onContextMenu={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  onMouseDown={[Function]}
                  onMouseLeave={[Function]}
                  onMouseUp={[Function]}
                  onTouchEnd={[Function]}
                  onTouchMove={[Function]}
                  onTouchStart={[Function]}
                  tabIndex="0"
                  type="button"
                >
                  <span
                    className="label-196"
                  >
                    <svg
                      aria-hidden="true"
                      className="root-186"
                      focusable="false"
                      role="presentation"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                      />
                      <path
                        d="M0 0h24v24H0z"
                        fill="none"
                      />
                    </svg>
                  </span>
                  <span
                    className="root-289"
                  />
                </button>
              </a>
            </div>
            <svg
              aria-hidden="true"
              className="root-186 titleIcon-5"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M0 0h24v24H0V0z"
                fill="none"
              />
              <path
                d="M20 6h-8l-2-2H4c-1.1 0-1.99.9-1.99 2L2 18c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V8c0-1.1-.9-2-2-2zm0 12H4V8h16v10z"
              />
            </svg>
             Scénarios
          </h4>
          <hr
            className="root-208"
          />
        </div>
        <div
          className="item-44 12-83"
        >
          <form
            autoComplete="off"
            className="formContainer-12"
            noValidate={true}
            onSubmit={[Function]}
          >
            <div
              className="container-43"
            >
              <div
                className="item-44 grow-7"
              >
                <div
                  className="root-213 marginNormal-214 fullWidth-216"
                >
                  <label
                    className="root-228 root-217 formControl-222 animated-225 outlined-227"
                    data-shrink={false}
                  >
                    Filtrer
                  </label>
                  <div
                    className="root-248 root-235 fullWidth-257 formControl-249"
                    onClick={[Function]}
                  >
                    <fieldset
                      aria-hidden={true}
                      className="root-265 notchedOutline-242"
                      style={
                        Object {
                          "paddingLeft": 8,
                        }
                      }
                    >
                      <legend
                        className="legend-266"
                        style={
                          Object {
                            "width": 0.01,
                          }
                        }
                      >
                        <span
                          dangerouslySetInnerHTML={
                            Object {
                              "__html": "&#8203;",
                            }
                          }
                        />
                      </legend>
                    </fieldset>
                    <input
                      aria-invalid={false}
                      className="input-258 input-243"
                      disabled={false}
                      onBlur={[Function]}
                      onChange={[Function]}
                      onFocus={[Function]}
                      required={false}
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </div>
              <div
                className="item-44 filterButtonsContainer-13"
              >
                <div
                  className="filterButtons-14"
                >
                  <button
                    className="root-205 root-267"
                    disabled={false}
                    onBlur={[Function]}
                    onContextMenu={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    onKeyUp={[Function]}
                    onMouseDown={[Function]}
                    onMouseLeave={[Function]}
                    onMouseUp={[Function]}
                    onTouchEnd={[Function]}
                    onTouchMove={[Function]}
                    onTouchStart={[Function]}
                    tabIndex="0"
                    type="submit"
                  >
                    <span
                      className="label-272"
                    >
                      <svg
                        aria-hidden="true"
                        className="root-186"
                        focusable="false"
                        role="presentation"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                        />
                        <path
                          d="M0 0h24v24H0z"
                          fill="none"
                        />
                      </svg>
                    </span>
                    <span
                      className="root-289"
                    />
                  </button>
                  <button
                    className="root-205 root-267"
                    disabled={false}
                    onBlur={[Function]}
                    onClick={[Function]}
                    onContextMenu={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    onKeyUp={[Function]}
                    onMouseDown={[Function]}
                    onMouseLeave={[Function]}
                    onMouseUp={[Function]}
                    onTouchEnd={[Function]}
                    onTouchMove={[Function]}
                    onTouchStart={[Function]}
                    tabIndex="0"
                    type="button"
                  >
                    <span
                      className="label-272"
                    >
                      <svg
                        aria-hidden="true"
                        className="root-186"
                        focusable="false"
                        role="presentation"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                        />
                        <path
                          d="M0 0h24v24H0z"
                          fill="none"
                        />
                      </svg>
                    </span>
                    <span
                      className="root-289"
                    />
                  </button>
                </div>
              </div>
            </div>
          </form>
        </div>
        <div
          className="item-44 12-83"
        >
          <div
            className="container-43 40-69"
          >
            <div
              className="item-44"
            >
              <p
                className="root-144 body1-153 gutterBottom-171"
              >
                Il n'y a pas de scénario à afficher
              </p>
            </div>
            <div
              className="item-44 12-83 hide-11"
            >
              <p
                className="root-144 body1-153 gutterBottom-171 loadMoreItem-10"
              >
                <a
                  className="root-144 colorPrimary-174 inline-179 root-180 underlineHover-182 loadMoreLink-9"
                  onClick={[Function]}
                >
                  Afficher plus de résultats
                </a>
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </main>
</div>
`;
